sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -102,56 +163,377 @@
-  public abstract static class Write extends PTransform<PCollection<Mutation>, PDone> {
+  public abstract static class Read extends PTransform<PBegin, PCollection<Struct>> {
+
+    private static final long serialVersionUID = 3124449388202574415L;
+
+    abstract SpannerConfig getSpannerConfig();
 
     @Nullable
-    abstract String getProjectId();
+    abstract TimestampBound getTimestampBound();
 
     @Nullable
-    abstract String getInstanceId();
+    abstract Statement getQuery();
 
     @Nullable
-    abstract String getDatabaseId();
+    abstract String getTable();
 
-    abstract long getBatchSizeBytes();
+    @Nullable
+    abstract String getIndex();
 
     @Nullable
-    @VisibleForTesting
-    abstract ServiceFactory<Spanner, SpannerOptions> getServiceFactory();
+    abstract List<String> getColumns();
+
+    @Nullable
+    abstract KeySet getKeySet();
+
+    @Nullable
+    abstract PCollectionView<Transaction> getTransaction();
 
     abstract Builder toBuilder();
 
+    public Read withTimestamp(Timestamp timestamp) {
+      return withTimestampBound(TimestampBound.ofReadTimestamp(timestamp));
+    }
+
+    public Read withTimestampBound(TimestampBound timestampBound) {
+      return toBuilder().setTimestampBound(timestampBound).build();
+    }
+
+    public Read withTable(String table) {
+      return toBuilder().setTable(table).build();
+    }
+
+    public Read withColumns(String... columns) {
+      return withColumns(Arrays.asList(columns));
+    }
+
+    public Read withColumns(List<String> columns) {
+      return toBuilder().setColumns(columns).build();
+    }
+
+    public Read withQuery(Statement statement) {
+      return toBuilder().setQuery(statement).build();
+    }
+
+    public Read withQuery(String sql) {
+      return withQuery(Statement.of(sql));
+    }
+
+    public Read withKeySet(KeySet keySet) {
+      return toBuilder().setKeySet(keySet).build();
+    }
+
+    public Read withIndex(String index) {
+      return toBuilder().setIndex(index).build();
+    }
+
+    public Read withSpannerConfig(SpannerConfig spannerConfig) {
+      return toBuilder().setSpannerConfig(spannerConfig).build();
+    }
+
     @AutoValue.Builder
     abstract static class Builder {
+      abstract SpannerConfig.Builder spannerConfigBuilder();
+
+      abstract Builder setSpannerConfig(SpannerConfig spannerConfig);
+
+      abstract Builder setTimestampBound(TimestampBound timestampBound);
+
+      abstract Builder setQuery(Statement statement);
+
+      abstract Builder setTable(String table);
+
+      abstract Builder setIndex(String index);
+
+      abstract Builder setColumns(List<String> columns);
+
+      abstract Builder setKeySet(KeySet keySet);
+
+      abstract Builder setTransaction(PCollectionView<Transaction> transaction);
+
+      abstract Read build();
+    }
+
+    /**
+     * Returns a new {@link SpannerIO.Read} that will write to the specified Cloud Spanner project.
+     *
+     * <p>Does not modify this object.
```

Comment:
Remove all the "Does not modify this object" comments - this is true for all withBlah() methods in Beam and specifying it is redundant.

You can also drop "Returns a new SpannerIO.Read that..." - usually we phrase these comments as simply "Specifies the Cloud Spanner project" (note: your comment currently says write - that's a copy-paste artifact I suppose; same is true for a bunch of other builder methods below)